### PR TITLE
feat: add CLI extended commands for lifecycle and observation

### PR DESCRIPTION
**Pull request by vicoooo26**
## Summary
- `pkg/cli/logs.go`: `arena job logs` — view pod logs
- `pkg/cli/delete.go`: `arena job delete` — delete jobs
- `pkg/cli/suspend.go`, `resume.go`: `arena job suspend/resume` — lifecycle management
- `pkg/cli/top.go`: `arena top job` — resource usage monitoring
- `pkg/cli/check.go`: `arena check` — environment and CRD verification
- `pkg/cli/integration_test.go`: integration tests (build-tagged)

Part of the Arena v2 CLI series (PR 6/7). Depends on #1465, #1466, #1467, #1468, #1469.

**Note:** Review only the last commit for this PR's changes.

## Test plan
- [ ] `go test ./pkg/cli/... -v` passes
- [ ] `go vet ./pkg/cli/...` passes
- [ ] `make arena-v2` produces binary with all commands
- [ ] `./bin/arena-v2 job --help` shows all subcommands

### CODE REVIEW

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
Thanks for the PR. The new lifecycle/observation commands follow the same patterns as the prior v2 CLI work, but I am requesting changes because of three merge blockers that need to be addressed first:

1. **go.mod / go.sum are out of sync.** The PR imports cobra, k8s.io/client-go, k8s.io/api, k8s.io/apimachinery, gopkg.in/yaml.v3, helm.sh/helm/v3, etc., but `go.mod` only declares `go 1.25.1` and `go.sum` is missing. A fresh `go build`/`go test` fails with missing modules. Please run `go mod tidy` and commit the results.

2. **E2E tests are missing.** The `v2-e2e-test` Makefile target points at `test/e2e/`, which does not exist. New CLI commands need e2e coverage against a real cluster.

3. **DCO is failing.** The DCO check is `ACTION_REQUIRED`; commits need `Signed-off-by` trailers. I already posted the fix commands in a separate comment.

I also left a few inline notes on selector construction, log streaming robustness, silent error swallowing in `top job`, and missing logs unit-test coverage.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	@mkdir -p test/e2e/crds
+	@echo "Downloading Kubeflow Training Operator CRDs..."
+	@curl -sL https://raw.githubusercontent.com/kubeflow/trainer/v1.9.3/manifests/base/crds/kubeflow.org_pytorchjobs.yaml -o test/e2e/crds/pytorchjobs.yaml
+	@curl -sL https://raw.githubusercontent.com/kubeflow/trainer/v1.9.3/manifests/base/crds/kubeflow.org_tfjobs.yaml -o test/e2e/crds/tfjobs.yaml
+	@echo "Downloading Kubeflow MPI Operator CRDs..."
+	@curl -sL https://raw.githubusercontent.com/kubeflow/mpi-operator/v0.8.0/manifests/base/kubeflow.org_mpijobs.yaml -o test/e2e/crds/mpijobs.yaml
+	@echo "CRDs downloaded to test/e2e/crds/"
+
+.PHONY: v2-e2e-test
+v2-e2e-test: v2-e2e-setup arena-v2 ## Run arena v2 e2e tests (requires real K8s cluster).
```
Comment (cheyang) [inline Makefile:98]: The `v2-e2e-test` target runs `go test -tags v2e2e ./test/e2e/`, but the PR does not include a `test/e2e/` directory. For a feature PR that adds new user-facing commands (delete, logs, suspend, resume, top, check), e2e coverage is required per the project bar. Please add e2e tests that exercise the lifecycle/observation commands against a real cluster, or escalate with a maintainer-approved plan if e2e is genuinely blocked.

File: pkg/provider/pytorch.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+// PyTorchProvider generates PyTorchJob CRDs.
+type PyTorchProvider struct{}
+
+func (p *PyTorchProvider) GetJobType() string {
+	return constants.KindPyTorchJob
+}
+
+func (p *PyTorchProvider) GetLogPodSelector(jobName string) string {
```
Comment (cheyang) [inline pkg/provider/pytorch.go:32]: GetLogPodSelector builds the label selector with `fmt.Sprintf` using the raw job name. Kubernetes resource names are constrained, so this is unlikely to be exploitable today, but it is fragile if the value ever comes from a less constrained source. Consider constructing selectors through `metav1.LabelSelector` + `metav1.FormatLabelSelector` so escaping is handled consistently. Same pattern appears in pkg/provider/tensorflow.go, pkg/provider/mpi.go, and the TFJob fallback in pkg/cli/logs.go.

File: pkg/cli/logs.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+		// Stream logs from the pod.
+		req := clientset.CoreV1().Pods(ns).GetLogs(podName, logOptions)
+		stream, err := req.Stream(cmdContext(cmd))
+		if err != nil {
+			return fmt.Errorf("failed to stream logs from pod %s: %w", podName, err)
+		}
+		defer stream.Close()
+
+		scanner := bufio.NewScanner(stream)
```
Comment (cheyang) [inline pkg/cli/logs.go:171]: The log stream is consumed with `bufio.Scanner`, which silently fails on lines longer than 64KB and returns `ErrTooLong`. For training job logs that can emit very long JSON lines or stack traces, this will cause the command to error out unexpectedly. Consider setting a larger scanner buffer or using `io.Copy`/`io.TeeReader` for robust streaming.

File: pkg/cli/top.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		ns := resolveNS("")
+		var allJobs []client.JobStatus
+		for _, kind := range supportedJobKinds {
+			if kind == constants.KindMPIJob && !mpiAvailable {
+				continue
+			}
+			jobs, err := k8sClient.List(cmdContext(cmd), kind, ns, V2LabelSelector)
+			if err != nil {
```
Comment (cheyang) [inline pkg/cli/top.go:60]: Listing failures for individual CRD kinds are logged with `log.Warning` and then silently continued. While this is reasonable when a CRD is simply not installed, it also masks permission errors or transient API failures. Consider distinguishing not-found errors from other API errors so users are not left thinking `top job` succeeded when one framework was silently skipped.

File: pkg/cli/logs_test.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
+package cli
```
Comment (cheyang) [inline pkg/cli/logs_test.go:1]: The logs tests only validate flag registration and kubeconfig failure. There is no test for pod selection logic, container validation, fallback to the TFJob worker selector, or the pod-belongs-to-job check. Consider adding fake-client tests that exercise these paths.

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
Thanks for the continued work on the arena v2 CLI. The unit tests pass and `go vet` is clean, but this PR still needs a couple of blockers addressed before it can merge:

1. **DCO sign-off is missing** on commits `33fec315` and `28c5337d`. The DCO bot is currently failing.
2. **No e2e tests under `test/e2e/`** despite the new `v2-e2e-test` Makefile target. For feature PRs adding lifecycle/observation commands we need real cluster coverage.

I also left a few non-blocking notes on log scanner robustness, `top job` error handling, and a question about `runPolicy.suspend` compatibility with MPIJob v1.

File: Makefile (hunk 1 above)
Comment (cheyang) [inline Makefile:98]: The `v2-e2e-test` target references `./test/e2e/`, but that directory does not exist in this PR. For a feature PR adding new CLI commands, we need real end-to-end coverage under `test/e2e/` that exercises the lifecycle and observation paths against a cluster. The Makefile target alone is not sufficient.

File: pkg/cli/logs.go (hunk 3 above)
Comment (cheyang) [inline pkg/cli/logs.go:171]: `bufio.NewScanner(stream)` uses the default 64KB token buffer. Training logs can emit very long JSON lines or stack traces; a single line over 64KB will cause `scanner.Err()` to return `bufio.ErrTooLong` and abort the stream.

File: pkg/cli/top.go (hunk 4 above)
Comment (cheyang) [inline pkg/cli/top.go:60]: When listing a CRD kind fails, the command logs a warning and continues. This can silently hide permission denials, network partitions, or a missing CRD, making `top job` report an incomplete job list without any error exit.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+v2-integration-test: ## Run arena v2 integration tests (no cluster required).
+	@echo "Running arena v2 integration tests..."
+	go test -tags integration ./pkg/cli/ -run TestIntegration -v -timeout 5m
+
+.PHONY: v2-install
+v2-install: ## Install arena v2 CLI to GOBIN.
+	@echo "Installing arena v2 CLI to $(GOBIN)..."
+	go install -ldflags '$(V2_LDFLAGS)' ./cmd/arena-v2/
+
+.PHONY: v2-e2e-setup
```
Comment (cheyang) [inline Makefile:84]: No e2e test package exists under `test/e2e/`. Please add Ginkgo-style tests that exercise `arena job delete/suspend/resume/logs/top/check` against a real cluster, even if they are tagged so they only run in CI.

File: pkg/cli/logs_test.go (hunk 5 above)
Comment (cheyang) [inline pkg/cli/logs_test.go:1]: `logs_test.go` only covers flag registration and arg validation. There is no test for provider selector construction, pod lookup, container validation, or log streaming behavior.

**Comment by cheyang** [comment]
@gemini-code-assist /review

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+v2-e2e-setup: ## Download Kubeflow Training Operator and MPI Operator CRDs for v2 e2e tests.
+	@mkdir -p test/e2e/crds
+	@echo "Downloading Kubeflow Training Operator CRDs..."
+	@curl -fsSL https://raw.githubusercontent.com/kubeflow/trainer/$(TRAINER_SHA)/manifests/base/crds/kubeflow.org_pytorchjobs.yaml -o test/e2e/crds/pytorchjobs.yaml
+	@curl -fsSL https://raw.githubusercontent.com/kubeflow/trainer/$(TRAINER_SHA)/manifests/base/crds/kubeflow.org_tfjobs.yaml -o test/e2e/crds/tfjobs.yaml
+	@echo "Downloading Kubeflow MPI Operator CRDs..."
+	@curl -fsSL https://raw.githubusercontent.com/kubeflow/mpi-operator/$(MPI_OPERATOR_SHA)/manifests/base/kubeflow.org_mpijobs.yaml -o test/e2e/crds/mpijobs.yaml
+	@echo "CRDs downloaded to test/e2e/crds/"
+
+.PHONY: v2-e2e-test
```
Comment (cheyang) [inline Makefile:97]: The new `v2-e2e-test` target is wired, but no files under `test/e2e/` are tagged with `//go:build v2e2e`, so `make v2-e2e-test` currently runs an empty suite. Because this PR introduces new user-facing commands (`job logs`, `job delete`, `job suspend`, `job resume`, `top job`, `check`), we need real e2e coverage that exercises them against a cluster. Please add the missing `test/e2e/` files and verify the target runs them.

File: go.mod [hunk 8]
```diff
@@ -1,3 +1,54 @@
...
+	github.com/json-iterator/go v1.1.12 // indirect
+	github.com/mailru/easyjson v0.7.7 // indirect
+	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
+	github.com/modern-go/reflect2 v1.0.3-0.20250322232337-35a7c28c31ee // indirect
+	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
+	github.com/pkg/errors v0.9.1 // indirect
+	github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2 // indirect
+	github.com/spf13/pflag v1.0.10 // indirect
+	github.com/x448/float16 v0.8.4 // indirect
+	go.yaml.in/yaml/v2 v2.4.4 // indirect
```
Comment (cheyang) [inline go.mod]: `go.yaml.in/yaml/v2` and `go.yaml.in/yaml/v3` look like typosquats of the legitimate `gopkg.in/yaml.v2`/`v3` packages. They are not imported anywhere in v2 code, yet they are listed as indirect requirements with checksums in `go.sum`. This is a supply-chain hygiene issue. Please run `go mod tidy` to remove unused modules and confirm the dependency graph only contains intended packages.

File: pkg/cli/logs.go (hunk 3 above)
Comment (cheyang) [inline pkg/cli/logs.go:171]: The log stream uses a default `bufio.Scanner`, which errors on lines larger than 64 KiB. Training job logs frequently exceed that (stack traces, tensor dumps). Consider calling `scanner.Buffer(make([]byte, 4096), maxTokenSize)` with a larger limit so streaming stays robust.

File: pkg/cli/top.go (hunk 4 above)
Comment (cheyang) [inline pkg/cli/top.go:60]: When listing a CRD kind fails, `top job` logs a warning and continues. On a partially broken cluster this can produce misleading output. Consider returning the error to the user or emitting a clear summary line that the results are partial.

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
Thanks for the extended CLI commands. This round still needs work before merge: the DCO check is failing because early commits use a GitHub noreply author email while the sign-off line uses a different address; no end-to-end tests exist for the new delete/suspend/resume/logs/top/check commands despite the Makefile target; and the v2 code path still depends on `helm.sh/helm/v3`, which the project guidelines prohibit. I also left non-blocking notes about log scanner robustness, top-job error handling, and logs test coverage.

File: Makefile (hunk 7 above)
Comment (cheyang) [inline Makefile:97]: The Makefile adds a `v2-e2e-test` target, but no test files exist under `test/e2e/`. For a feature PR that introduces `delete`, `suspend`, `resume`, `logs`, `top`, and `check` commands, we need end-to-end coverage (dry-run, envtest, or real-cluster Ginkgo tests) that exercises the new command wiring. Please add tests under `test/e2e/` and wire them to the existing make target.

File: pkg/task/setter.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
+package task
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+	"helm.sh/helm/v3/pkg/strvals"
```
Comment (cheyang) [inline pkg/task/setter.go]: `pkg/task/setter.go` imports `helm.sh/helm/v3/pkg/strvals`, and `go.mod` declares `helm.sh/helm/v3` as a direct dependency. The project instructions in `AGENTS.md`/`CLAUDE.md` explicitly forbid adding Helm as a v2 dependency. Please replace the Helm strvals parser with a self-contained implementation or an allowed, smaller library.

File: pkg/cli/logs.go (hunk 3 above)
Comment (cheyang) [inline pkg/cli/logs.go:171]: `bufio.NewScanner(stream)` uses the default 64 KiB token limit. Training logs can contain stack traces or JSON lines longer than that, causing the scanner to return `bufio.ErrTooToken` and surface an error to the user. Consider using a larger buffer (`scanner.Buffer`) or switching to `io.Copy`/`bufio.Reader` for robust streaming.

File: pkg/cli/top.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+			mpiAvailable = false
+		}
+
+		ns := resolveNS("")
+		var allJobs []client.JobStatus
+		for _, kind := range supportedJobKinds {
+			if kind == constants.KindMPIJob && !mpiAvailable {
+				continue
+			}
+			jobs, err := k8sClient.List(cmdContext(cmd), kind, ns, V2LabelSelector)
```
Comment (cheyang) [inline pkg/cli/top.go:59]: When `k8sClient.List` fails, the command logs a warning and continues, eventually rendering `No jobs found`. In a cluster with transient RBAC or network errors this can mislead users into thinking no jobs exist. Consider returning the error or exiting non-zero with a clear message when all list attempts fail.

File: pkg/cli/logs_test.go (hunk 5 above)
Comment (cheyang) [inline pkg/cli/logs_test.go:1]: `logs_test.go` covers flag registration and kubeconfig failures, but does not test provider selector fallback, container validation, pod ownership check, or log streaming behavior. Please add unit tests using fake clients or captured streams.

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
@vicoooo26 — this PR is blocked from merging by a **failing DCO check**.
The DCO bot requires every commit to carry a `Signed-off-by:` trailer; without it the PR cannot land regardless of `lgtm` / `approved` labels or otherwise-green CI.

Please resolve this now. For a single-commit PR:

```
git commit --amend -s
git push --force-with-lease
```

For multiple commits on this branch:

```
git rebase --signoff origin/develop-v2
git push --force-with-lease
```

The code review side is finished and any merge labels already on the PR will take effect once DCO turns green. Reference: https://github.com/apps/dco

**Comment by cheyang** [comment]
@gemini-code-assist /review
